[wanja] - reset.css 추가, VIew의 미들웨어 추가, 캐러셀 기능 구현, 프로그레스바 구현

wan0514 opened the pull request:
## 작업 완료 내역

### ✨ 기능 추가
- **ListView 및 GridView 데이터 및 상태 관리용 미들웨어 및 컨테이너 레이어 추가**
  - `ListView`와 `GridView`의 데이터 파싱 및 상태 관리를 위한 미들웨어와 컨테이너 레이어를 추가
  - `ListView`와 `GridView`의 상태 관리 및 데이터 처리를 미들웨어와 컨테이너 컴포넌트에서 담당하게 함으로써, 각 컴포넌트는 프레젠테이션에 집중하도록 개선
  - 데이터 구조와 렌더링 상태 처리를 미들웨어에서 관리하여 코드의 유지 보수성과 재사용성 향상

- **GridView와 ListView에서 `useCarousel` 훅 사용**
  - `MediaSection`에서 사용된 공통 로직을 `useCarousel` 훅으로 이동시켜 재사용 가능하도록 함.

- **공용 Carousel 버튼 및 컨테이너 컴포넌트 추출**
  - Carousel 버튼 및 컨테이너 로직을 `CarouselContainer`로 추출
  - `GridViewContainer`와 `ListViewContainer`에서 공통 로직을 재사용하여 간소화

- **Progress Bar 구현**
  - 프로그레스바 컴포넌트를 새로 생성하여, `TabButton`의 하위 레이어에 배치
  - `TabButton`이 활성화되었을 때만 프로그레스바가 나타나도록 규현


### 🐛 버그 수정
- **카테고리 변경 시 ListView 렌더링 오류 수정**
  - `currentPage` 상태 업데이트 전 이전 페이지 값으로 렌더링되는 문제 해결
  - 카테고리 변경과 `currentPage` 상태 업데이트 동기화 문제 해결

---

### 💄 스타일 설정
- **`reset.css` 적용하여 스타일 초기화**
  - 기존의 Emotion reset으로 충분하지 않아서 `reset.css`를 추가하여 스타일 초기화 개선

- **레이아웃 높이 고정 및 테두리 스타일 수정**
  - 레이아웃 높이를 고정하여 리스트 데이터에 따라 높이가 변하지 않도록 설정
  - 고정된 높이에 맞게 테두리 스타일을 업데이트하여 일관된 레이아웃 유지

---

### 🏗️ 구조 및 설계
- **캐러셀 네비게이션 구현 (`useCarousel` 훅 사용)**
  - `useCarousel` 훅을 사용하여 수동 및 자동 캐러셀 기능 구현

- **카테고리별 첫/마지막 페이지 네비게이션 기능 추가**
  - 카로셀에서 첫 페이지에서 이전 카테고리의 마지막 페이지로, 마지막 페이지에서 다음 카테고리의 첫 페이지로 이동하는 기능 추가
  - `fieldMap`을 상수 파일로 분리하여 카테고리 네비게이션 지원
  - `reset` 함수에 페이지 번호 매개변수 추가하여 특정 페이지로 직접 이동할 수 있도록 확장

- **리스트 페이지 자동 페이지 전환 기능 (20초 간격)**
  - 20초마다 자동으로 다음 페이지로 전환하는 기능 구현

---

### 주요 고민 및 해결과정

### 0. **뷰 컴포넌트의 역할 분리를 위해 데이터를 어디서 관리할 것인가?**

#### 🧐 **고민의 배경**

- **기존 구조**:
  - 기존에는 **미들웨어** 없이 **`ListView`**와 **`GridView`**가 **데이터를 직접 받아 처리**하고, **페이지네이션**과 **카테고리 상태**를 모두 **뷰 컴포넌트**에서 관리하고 있었습니다.
  - 이 구조는 **뷰의 책임이 커지며** 컴포넌트 간 **책임 분리가 불명확**해졌습니다.
  - **상태 관리**와 **데이터 처리**가 뷰 컴포넌트에 집중되었고, 이는 **코드 중복**을 초래하고 **유지보수**가 어려운 문제를 발생시켰습니다.

#### 2. **고려한 방식**

- **미들웨어 역할을 하는 컨테이너 컴포넌트 생성**:
  - **`ListViewContainer`**와 **`GridViewContainer`** 각각을 생성하여, **데이터와 상태**를 관리하도록 했습니다.
  - **뷰 컴포넌트**는 이제 **UI 렌더링**에만 집중하고, 데이터와 상태 관리 작업은 **컨테이너 컴포넌트**에서 처리합니다.
  
- **상위 컨테이너에서 데이터와 상태 관리**:
  - 각 **뷰 컴포넌트**는 **필요한 데이터만** 받게 하고, 데이터를 **정제하고 상태를 관리하는 역할**은 **컨테이너 컴포넌트**가 담당하게 했습니다.
  - **`GridViewContainer`**와 **`ListViewContainer`**는 **데이터를 처리**하고, **뷰 컴포넌트**는 **렌더링**만을 담당하도록 **역할을 분리**했습니다.

- **공통된 로직을 하나의 컴포넌트로 추출**:
  - **뷰 컴포넌트**는 이제 **`GridViewContainer`** 또는 **`ListViewContainer`**와 같은 **전용 컨테이너**를 통해 **데이터와 상태를 전달**받고, **렌더링에만 집중**할 수 있게 되었습니다.

#### 3. **최종 선택과 이유**

- **책임 분리**:
  - 각 컴포넌트가 **자신의 책임을 명확히 하도록** 변경했습니다. **뷰 컴포넌트**는 **UI 렌더링**만 담당하고, **데이터 처리 및 상태 관리**는 **컨테이너 컴포넌트**에서 처리하도록 했습니다.
  - 이 방식은 **단일 책임 원칙**을 따르며, **유지보수**와 **확장성**을 높이는 데 기여했습니다.

- **중복 코드 제거**:
  - **`ListViewContainer`**와 **`GridViewContainer`**에서 **데이터 처리 로직**을 각각 관리하면서, **뷰 컴포넌트**는 **렌더링 역할**만 하게 되어 **코드 중복**을 제거하고 **관리 용이성**이 향상되었습니다.
  
- **확장성**:
  - 새로운 뷰 컴포넌트가必要할 때, **`GridViewContainer`**와 **`ListViewContainer`**와 같은 **전용 컨테이너**를 **재사용**할 수 있어, **확장성이 뛰어난 구조**가 되었습니다.
  - **페이지네이션**이나 **기타 기능**을 추가할 때도, **컨테이너 컴포넌트**에서 관리하면 되므로, **확장에 용이**합니다.

---


### 1. **캐러셀 버튼 로직을 어떻게 공통으로 쓸까?**

#### 🧐 **고민의 배경**

- **리스트**와 **그리드**에서 **공통적인 캐러셀 버튼**을 사용하려고 했으나, **동작의 차이**로 인해 문제가 발생했습니다.
- **리스트**는 **무한 캐러셀과 자동 페이지 전환**이 필요했고, **그리드**는 **수동 페이지 전환**만 필요했습니다.
- **목표**는: **리스트**와 **그리드**의 특성을 고려하면서도 공통 로직을 효율적으로 처리할 수 있는 방법을 찾는 것이었습니다.

#### **고려한 방식**

1. **`useCarousel`에서 로직 분기 처리**  
   - `viewType`을 기반으로 **리스트**와 **그리드**의 **동작을 분기**하는 방법을 고려했습니다.  
   - **리스트**와 **그리드**에서 동작이 다르므로, **`useCarousel` 훅** 내에서 **조건에 따라** 자동/수동 전환 로직을 처리하려 했습니다.

2. **`useCarousel`은 공통 로직만 두고, 외부에서 추가 로직 정의**  
   - **공통 로직**은 **`useCarousel`**에 두고, 각 **리스트**와 **그리드**에서 필요한 추가적인 **로직**을 **각각의 뷰에서 처리**하도록 하여 **유지보수성**을 높이기로 결정했습니다.  
   - **핸들러(`goNext`, `goPrev`)**는 **리스트/그리드**에서 각각 정의하고, **`useCarousel`**에서 공통된 기능을 관리하는 방식을 택했습니다.

#### ✅ **최종 해결**

- **`useCarousel`**은 **공통 로직만 처리**하고, 각 뷰 컴포넌트에서 필요한 **추가 로직을 정의**하는 방식으로 처리하기로 했습니다.
- **리스트**와 **그리드에서 필요한 로직**을 각 컴포넌트에서 독립적으로 관리하여, **동작이 다르더라도 공통 훅을 활용**할 수 있게 됐습니다.
- **`MediaSection`**에서는 공통적으로 **`useCarousel`**을 사용하고, **리스트/그리드에서 필요한 로직만 별도로 처리**하여 **상위 컴포넌트의 역할을 최소화**했습니다.

---

### 2. **각 카테고리의 마지막 페이지일 경우 다음 카테고리로 넘어가게 할까?**

#### 🧐 **고민의 배경**

- **리스트**에서 **각 카테고리의 마지막 페이지**에 도달했을 때, **다음 카테고리로 자동 전환**하는 로직을 구현해야 했습니다.
- **그리드**에서는 **카테고리 전환이 필요 없고**, **리스트**에서만 **카테고리 전환**이 필요했습니다.  
- 이를 **어떻게 처리할지**에 대한 고민이 있었습니다.

#### **고려한 방식**

1. **`useCarousel`에서 카테고리 상태를 내부에서 처리**  
   - **`useCarousel` 훅**에서 **카테고리 상태**를 관리하고, **페이지 이동**과 **카테고리 전환**을 내부에서 처리하는 방법을 고려했습니다.

2. **`useCarousel`에서 기본적인 페이지 상태만 관리하고, 외부에서 `goNext`와 같은 함수로 동작 구현**  
   - **리스트 컴포넌트에서 `goNext`, `goPrev`와 같은 핸들러를 정의**하고, **캐러셀 버튼에 해당 핸들러를 전달**하는 방식으로, **리스트에서 캐러셀 버튼을 클릭했을 때 동작**을 **한눈에 보이게 처리**하려는 방식이 더 직관적이라고 판단했습니다.

#### ✅ **최종 선택**

- **`useCarousel`**은 **공통 로직만 처리**하고, **핸들러(`goNext`, `goPrev`)는 외부에서 정의**하여 **리스트와 그리드에서 동작을 분리**했습니다.
- **핸들러**는 **리스트**에서만 카테고리 전환 로직을 처리하도록 했고, **`useCarousel` 훅은 페이지 상태 관리만 맡게 했습니다.
- **핸들러를 각 뷰에서 정의**하고 **캐러셀 버튼에 전달**하는 방식이 더 **가독성이 높고, 코드 흐름을 쉽게 파악**할 수 있다고 판단했습니다.

#### **선택의 근거**

1. **코드 읽는 흐름을 직관적으로 만들기 위해**  
   - **`useCarousel`**에 **카테고리 전환 로직**을 넣으면, 코드가 **복잡해지고 이해하기 어려운** 구조가 될 것이라고 생각했습니다.  
   - 대신 각 뷰에서 **핸들러를 정의**하고, **각 동작을 한 곳에서 처리**하는 방식이 더 **직관적이고 유지보수에 용이**하다고 판단했습니다.

2. **유지보수성과 확장성**  
   - `useCarousel`은 **단순한 페이지 이동 로직**만 처리하고, **각 뷰에서 필요한 로직**을 **독립적으로 관리**하여 **확장성**이 높고 **재사용성이 뛰어나**지도록 했습니다.

---

### 3. **캐러셀 버튼, `useCarousel`, `totalPage` 상태를 어디에 정의할 것인가?**

#### 🧐 **고민의 배경**

처음에는 **`MediaSection`**에서 **캐러셀 버튼**과 **`useCarousel` 훅**을 모두 정의하고, 그 안에서 **리스트**와 **그리드** 컴포넌트로 **상태를 전달**하는 방식으로 접근했습니다.  
이 방식은 **상위 컴포넌트에서 모든 상태를 관리**할 수 있다는 장점이 있지만, **리스트**와 **그리드**의 **동작이 다르기 때문에** **동작을 분기 처리**하는 데 어려움이 있었습니다.

#### **고민한 점과 고려한 사항**

1. **상위 컴포넌트에서 관리**할 경우
   - **`MediaSection`에서** **`useCarousel`**과 **캐러셀 버튼**을 **모두 정의**하고, **상위 컴포넌트에서 상태 관리**를 하면 **공통적으로 사용**할 수 있다는 장점이 있었습니다.
   - 하지만 **리스트**와 **그리드**에서 **동작이 다르기** 때문에, **상위 컴포넌트에서 분기 처리**해야 하므로 **불필요하게 복잡해질 수 있다는** 단점이 있었습니다.

2. **리스트와 그리드에서 각자 관리**할 경우
   - **`totalPage`**와 **`useCarousel`**을 **각각의 컨테이너에서 관리**하고, **각각의 뷰에서 필요한 로직을 처리**하면 **유지보수성**이 더 좋고, **리스트와 그리드의 요구사항을 명확히 구분**할 수 있습니다.
   - 이 방식은 **상위 컴포넌트에서 역할을 최소화**하고, **각 뷰에서 독립적으로 동작을 관리**하는 방식으로 유리하다고 판단했습니다.

#### ✅ **최종 선택**

- **`useCarousel`과 캐러셀 버튼을** **각각의 리스트/그리드 컨테이너에서 정의**하고, **상위 컴포넌트(`MediaSection`)에서는 뷰만 관리**하도록 결정했습니다.
- **`totalPage` 상태**는 **리스트와 그리드에서** **각각의 필요에 맞게 계산하고 관리**하도록 했습니다. 이렇게 하면 **상위 컴포넌트**에서는 **뷰만 관리**하고, **각 뷰에서 필요한 상태**를 **직접 관리**할 수 있습니다.

---

### 5. **재사용성을 위한 리팩토링: 공통 캐러셀 컨테이너 컴포넌트 추출**

#### 🧐 **고민의 배경**

프로젝트를 진행하면서 `GridViewContainer`와 `ListViewContainer` 컴포넌트는 매우 유사한 구조를 가지고 있었습니다. 두 컴포넌트는 공통적으로 **캐러셀 버튼** (이전, 다음 버튼)과 **컨테이너** (미디어 컨테이너)를 사용하고 있었습니다. 하지만 이들 각각의 컴포넌트에서 버튼과 컨테이너의 로직을 중복해서 작성하고 있었고, 이는 코드 중복을 초래하고 유지보수가 어려운 문제를 발생시켰습니다. 

### **문제점:**
1. **중복 코드**: `GridViewContainer`와 `ListViewContainer`에서 캐러셀 버튼과 미디어 컨테이너 로직을 반복적으로 작성하고 있었기 때문에, 코드의 중복성이 높아지고 수정이 필요할 경우 모든 컴포넌트를 찾아 수정해야 했습니다.
2. **유지보수 어려움**: 각 뷰에서 캐러셀 버튼과 컨테이너 로직을 수정하거나 확장해야 할 경우, 각 컴포넌트에서 동일한 수정 작업을 해야 하므로 실수할 가능성도 높아지고 유지보수가 복잡해졌습니다.
3. **확장성 부족**: 새로운 뷰를 추가할 때마다 동일한 구조를 다시 작성해야 하기 때문에, 코드의 확장성도 제한적이었습니다.

#### ✅ **문제 해결 과정**

이 문제를 해결하기 위해 **공통된 로직을 하나의 컴포넌트로 추출하는 방식**을 택했습니다. `CarouselContainer`라는 컴포넌트를 새로 만들어, **캐러셀 버튼**과 **미디어 컨테이너**의 공통 로직을 한 곳에서 처리하게 했습니다. 이 방식은 여러 뷰 컴포넌트에서 공통적으로 사용될 수 있는 로직을 재사용 가능하도록 만들었습니다.

### **설계 방식:**
1. **`CarouselContainer` 컴포넌트 생성**: `CarouselContainer` 컴포넌트를 새로 만들어, 버튼과 미디어 컨테이너의 공통 로직을 처리하도록 했습니다. 이 컴포넌트는 `children`으로 전달되는 뷰를 감싸는 역할만 하며, **버튼 클릭 시의 이벤트 처리**와 **버튼의 표시 여부**를 제어합니다. 버튼 클릭 시에는 `goPrev`와 `goNext`와 같은 페이지 전환 로직을 처리하는 함수들을 외부에서 전달받습니다.
   
2. **조건부 버튼 렌더링**: `CarouselContainer`는 `showPrev`와 `showNext` props를 사용하여, 각 뷰에서 페이지 전환에 필요한 버튼을 **조건부로 렌더링**합니다. 예를 들어, 그리드 뷰에서는 `prev` 버튼을 표시하고, 리스트 뷰에서는 버튼을 숨기거나 필요한 경우에만 렌더링합니다.

3. **뷰 컴포넌트에서 공통 로직 사용**: 이제 `GridViewContainer`와 `ListViewContainer`는 공통된 `CarouselContainer`를 사용하여 코드 중복을 제거했습니다. 각 뷰는 버튼의 표시 여부와 페이지 전환 로직만 설정하고, 나머지 로직은 `CarouselContainer`에서 처리하게 되었습니다. 이로 인해 각 컴포넌트는 더 깔끔하고 관리하기 쉬워졌습니다.

4. **확장성 고려**: 새로운 뷰 컴포넌트를 추가할 때마다, 해당 뷰는 `CarouselContainer`를 사용하여 버튼과 컨테이너의 공통 로직을 재사용할 수 있게 되었습니다. 이로써 새로운 뷰를 추가하는 데 드는 비용이 크게 줄어들었습니다.

---
 ## 4/9: : progress bar 관련 문제

### 문제 6: **프로그레스바 관련 로직을 상위에 둘 것인가, 내부에 둘 것인가?**

#### **🤔 고민의 배경**
프로그레스바를 구현하는 과정에서 **로직을 어디에 배치할지**에 대한 고민이 있었습니다. `ProgressBar`는 **진행 상태(progress)**에 따라 **UI를 업데이트**하는 컴포넌트입니다. 하지만 이 컴포넌트가 **상태를 관리하고 애니메이션 로직을 처리**할지, 아니면 **상위 컴포넌트에서 상태와 로직을 처리하고** `ProgressBar`는 **UI만 담당**할지에 대한 결정이 필요했습니다. 고민의 핵심은 **책임의 분리**와 **구성의 단순화**였습니다. 

#### **🤔 고려한 방식 두 가지**

##### **1. 프로그레스바 로직을 내부에 두는 방식**
- `ProgressBar` 내부에서 진행률을 계산하고 애니메이션을 처리하는 방식입니다.
- UI와 로직이 **`ProgressBar` 안에 모두 포함**됩니다.

###### **장점**
- **단순화**: 코드가 간결하고 직관적입니다.
- **캡슐화**: `ProgressBar`가 **UI와 로직을 모두 처리**하여 다른 컴포넌트에서 쉽게 사용할 수 있습니다.
- **빠른 구현**: 복잡한 상태 관리 없이 UI와 애니메이션을 **빠르게 구현**할 수 있습니다.

###### **단점**
- **유연성 부족**: 로직을 확장하거나 변경할 때 복잡해질 수 있습니다.
- **책임 복잡**: UI와 로직이 결합되어 컴포넌트가 커지거나 복잡해질 수 있습니다.
- **상위 컴포넌트와의 의존성**: 상위에서 상태 제어가 어려워집니다.

##### **2. 프로그레스바 로직을 외부로 빼는 방식**
- `ProgressBar`는 UI만 담당하고, 상위 컴포넌트에서 상태 관리와 애니메이션 로직을 처리하는 방식입니다.

###### **장점**
- **책임 분리**: UI와 로직을 분리하여 각각의 역할을 명확히 할 수 있습니다.
- **유연성 향상**: 로직을 상위 컴포넌트에서 처리하므로, 상태 관리나 **페이지 전환** 등을 더 유연하게 처리할 수 있습니다.
- **확장성**: 다른 상태나 로직을 연동하거나 다른 애니메이션 효과를 적용할 때 더 유연하게 처리할 수 있습니다.
- **UI만 담당**: 다른 UI 요소에 대한 처리가 필요할 때 쉽게 관리할 수 있습니다.

###### **단점**
- **상위 컴포넌트 복잡도 증가**: 상위 컴포넌트가 복잡해질 수 있습니다.
- **UI와 로직 연결 필요**: UI와 로직 간의 연결이 필요하고 관리가 번거로울 수 있습니다.

##### **✅ 최종 판단과 선택**
- **프로그레스바는 UI만 담당**해야 한다는 원칙을 바탕으로, 로직은 **상위 컴포넌트에서 처리**하는 방식으로 선택하였습니다.
- 이는 **책임 분리**와 **유연성**을 고려한 결정이었습니다.
- **상위 컴포넌트에서 로직을 처리**하면 **확장성과 유연성**이 더 용이하고, **UI는 `ProgressBar`에 집중**할 수 있습니다.

>cf: 개발을 하면서 ListContainer(상위 컴포넌트)의 역할과 상태 관리가 점점 많아짐을 느끼고 있습니다. 내부에 넣는 방식도 고려중이지만, 구현을 우선적으로 하는 중이라 추후 리팩토링 과정에서 시도해볼 예정입니다.

---

### 문제 7: **Progress Bar가 페이지 변경 시 동작하지 않음**
#### 🤔 배경
- 페이지 변경 시 프로그레스바가 동기화되지 않거나, 20초보다 빠르게 차오르는 문제가 발생했습니다.

#### ✅ 해결: **동기화된 Progress Bar 동작 구현**
- 페이지 변경과 프로그레스바 업데이트를 각각 독립적으로 관리하던 `setInterval`을 하나로 통합하여 코드 중복을 줄였습니다.
- 20초 동안 차오르는 시간을 정확하게 맞추기 위해 진척도를 계산하는 로직을 수정했습니다.
- **duration** 상수를 도입하여 시간 변경 시, **`duration`만 수정하면** 다른 시간으로도 쉽게 조정 가능하도록 했습니다.
- 페이지 변경 시마다 프로그레스바는 0%로 초기화되며, 설정된 시간 동안 차오르도록 동기화하여 동작하게 했습니다.

wan0514 committed:
♻️ refactor: add middleware and state for ListView

- introduced middleware component for ListView data parsing and state management
- improved handling of data structure and rendering state within the middleware layer

wan0514 committed:
♻️ refactor: clarify state and variable names, remove unused vars

- improved readability by renaming states and variables clearly
- removed unnecessary variables to clean up the codebase

wan0514 committed:
♻️ refactor: add container layer to GridView for data and state

- introduced container component to manage data and state for GridView
- simplified GridView to focus solely on presentation

wan0514 committed:
✨ feat: add carousel buttons and place in media area

- Create prev/next carousel button component
- Place carousel buttons in the media area
- Add test handler function
- Button functionality not implemented yet (test setup)

wan0514 committed:
✨ feat: implement manual carousel using useCarousel hook

- Implemented manual carousel navigation using custom hook useCarousel
- Auto-play feature is planned for future implementation
- Navigation to the first page of next category is not yet implemented

wan0514 committed:
💄 style: apply reset.css to enhance style reset

- Added reset.css as emotion reset was not sufficient
- Set up additional reset for clearer style initialization

wan0514 committed:
💄 style: fix layout height and adjust border styling

- Fixed layout height to prevent changes based on list data
- Ensured consistent layout with fixed height
- Updated border styling to match the fixed height

wan0514 committed:
🐛 fix: Fix listView render error when changing category

- Resolved issue where the listView was rendered with the previous currentPage value before the currentPage state was updated.
- Ensured proper rendering by synchronizing category change and currentPage state update.

wan0514 committed:
♻️ refactor: Use useCarousel hook in Grid and List

- Moved common logic previously used in MediaSection to the useCarousel hook

wan0514 committed:
✨feat: Add category navigation for first/last page in list carousel

- Added functionality to navigate to the last page of the previous category when at the first page.
- Added functionality to navigate to the first page of the next category when at the last page.
- Separated `fieldMap` into a constant file to support category navigation.
- Extended the `reset` function to accept a page number parameter, enabling direct navigation to a specific page.

wan0514 committed:
✨feat: Add auto-page navigation every 20 seconds in the list page

- Implemented a feature to automatically navigate to the next page every 20 seconds.

wan0514 committed:
♻️ refactor: Remove unnecessary parameters and functions in useCarousel

wan0514 committed:
♻️ refactor: Extract common Carousel button and container components

- Extracted common button and container logic into CarouselContainer
- Simplified GridViewContainer and ListViewContainer by using the shared CarouselContainer

wan0514 committed:
♻️ refactor: Extract common logic for calculating category index

- Extracted category index calculation logic into a reusable function

wan0514 committed:
♻️ refactor: Rename variables, functions, props, and states for clarity

- Renamed variables, functions, props, and states to more intuitive and readable names
- Improved code readability and understanding by using more descriptive names

wan0514 committed:
✨ feat: Implement progress bar for category buttons

- Created a progress bar component and applied it only to the active category buttons.
- Used setInterval to synchronize the page transition every 20 seconds with the progress bar.
- Currently, setInterval is used independently for both, so synchronization will be needed in the future.

wan0514 committed:
♻️  refactor: Synchronize page auto-change and progress bar update

- Combined page auto-change and progress bar update into a single setInterval

wan0514 committed:
♻️ refactor: Refactor progress bar update logic by separating duration constant

- Extracted the duration value into a constant for better readability and maintainability

crongro commented:
> * Carousel 버튼 및 컨테이너 로직을 `CarouselContainer`로 추출
> * `GridViewContainer`와 `ListViewContainer`에서 공통 로직을 재사용하여 간소화

공통부분을 분리한것 좋네요.
어떤 경우에는 이런 상황에서도 커스텀훅이 해결책이 될 수도 있고요.

crongro commented on src/components/Media/MediaSection/Carousel/useCarousel.js:
```diff
@@ -0,0 +1,24 @@
+export default function useCarousel() {
+  const [currentPage, setCurrentPage] = useState(0);
+
+  function goNext() {
+    setCurrentPage((prev) => prev + 1);
+  }
+
+  function goPrev() {
+    setCurrentPage((prev) => prev - 1);
+  }
```
한개 함수로 만들수도 있을듯.
direction (next/prev)를 문자열 파라미터로 넘겨주어서.

crongro commented on src/components/Media/MediaSection/ListView/index.jsx:
```diff
@@ -1,83 +1,105 @@
+
+const ListViewContainer = ({ data }) => {
+  const [totalPage, setTotalPage] = useState(0);
+  const { currentPage, goNext, goPrev, reset } = useCarousel();
```
커스텀훅 분리하고 재사용이 되는건가요? 좋은 시도군요. 
저 커스텀훅도 파라미터로 초기에 값을 전달할 것이 있다면 전달해도 되고요.

crongro commented on src/components/Media/MediaSection/ListView/index.jsx:
```diff
@@ -1,83 +1,105 @@
+  const pressList = data[currentCategory];
+  const selectedPressData = pressList[currentPage];
+
+  const getCurrentCategoryIndex = useCallback(
```
함수 참조를 유지하는것인데, 그 효과는 어떨지도 생각해보세요

crongro commented on src/components/Media/MediaSection/ListView/index.jsx:
```diff
@@ -1,83 +1,105 @@
+  );
+
+  function goToPrevPage() {
+    if (currentPage === 0) {
```
들여쓰기는 항상 줄이기

if (currentPage !== 0) {
  goPrev();
  return;
}

wan0514 committed:
✨ feat : Add 'limit' query parameter to /news/grid API request

- Updated the API call to support 'limit' query parameter for item count control.
- The 'limit' parameter is now passed in the GET request to control the number of grid items.

crongro commented on src/components/Media/MediaSection/ListView/index.jsx:
```diff
@@ -1,83 +1,105 @@
+        }
+        return prev + 100 / (DURATION / 100);
+      });
+    }, 100); // 100ms마다 진행률 업데이트
```
100도 변수로 저장해두지 그래요

crongro commented on src/components/Media/MediaSection/ListView/index.jsx:
```diff
@@ -1,83 +1,105 @@
+    }, 100); // 100ms마다 진행률 업데이트
+
+    return () => {
+      clearInterval(pageChangeInterval);
+      clearInterval(progressInterval);
```
두 개지만..

[pageChangeInterval, progressInterval].forEach(clearInterval);

wan0514 replied:
> 두 개지만..
> 
> [pageChangeInterval, progressInterval].forEach(clearInterval);

😮 저런 방법도 있군요....... 유레카네요.....

crongro reviewed (approved):
문제점 개선 과정의 글이 좋네요

wan0514 committed:
✨ feat : Replace mock data with server API endpoint in fetch request

- Updated fetch logic to use the server's API endpoint instead of mock data.
- The API call now fetches real data from the server.

wan0514 committed:
♻️ refactor: Remove redundant code due to changed data from backend

- Backend data format has changed, leading to unnecessary code.
- Removed the obsolete code to improve clarity and maintainability.

wan0514 committed:
✨ feat: Create ContextAPI for managing subscribed press

wan0514 committed:
♻️ refactor: Rename media states and make TabButton component more flexible

- Renamed states for all media outlets and subscribed media outlets for clarity.
- Refactored TabButton component to determine label names internally based on tab type, improving flexibility.

wan0514 committed:
✨ feat: Apply ContextAPI to badge displaying number of subscribed media outlets

- Replaced hardcoded badge number with ContextAPI to dynamically reflect the count of subscribed media outlets.

wan0514 committed:
🐛 fix: Resolve overlapping border issue in TabList component

- Fixed the issue where the borders of TabList components were overlapping.
- Adjusted the CSS to prevent the double border effect.

crongro commented on src/components/Media/MediaTab/TabList/TabItem.jsx:
```diff
@@ -23,13 +23,22 @@ const LabelWrapper = styled.span`
         <LabelWrapper>
           {label}
-          {badgeLabel && <Badge label={badgeLabel} isActive={isActive} />}
+          {typeof badgeLabel === 'number' && (
```
숫자가 아닌 경우가 생겨요? 타입체크를 갑자기 해서 궁금했네요.

wan0514 replied:
> 숫자가 아닌 경우가 생겨요? 타입체크를 갑자기 해서 궁금했네요.

숫자가 아닌 경우가 생기지는 않지만, badgeLabel이 0일 경우 Badge 컴포넌트가 랜더링이 안되는 문제가 발생했습니다.
이전 코드에서 badgeLabel && <Badge ... />  에서 만약 badgeLabel이 0일 경우에는 falsy값으로 처리되어서 Badge가 랜더가 안되는 이슈가 있었습니다.

badgeLabel로 넘기는 값을 string으로 변환하는 방식과, TabItem내부적으로 badgeLabel을 string으로 바꾸거나, 0만 예외적으로 처리하는 등의 방법 중에서 고민해보았는데... 제가 생각하기에 뱃지에 숫자만 들어오니까 저 방식이 깔끔하다고 생각했습니다.

crongro commented on src/components/Media/index.jsx:
```diff
@@ -11,17 +11,23 @@ const MediaContainer = styled.div`
     Promise.all([
-      fetch('/mockData/pressList.json').then((res) => res.json()),
-      fetch('/mockData/mockData.json').then((res) => res.json()),
+      fetch(`http://localhost:5173/news/grid?limit=${LIMIT}`).then((res) =>
```
다음엔 로컬 URL도 환경설정으로 분리해볼래요?

crongro commented on vite.config.js:
```diff
@@ -21,4 +21,10 @@ export default defineConfig({
   },
+
+  server: {
+    proxy: {
+      '/news': 'http://localhost:3000', // Express 서버가 실행 중인 포트로 프록시 설정
+    },
```
이건 왜 사용해요? 어떤 문제때문에?

wan0514 replied:
> 이건 왜 사용해요? 어떤 문제때문에?

백엔드로 서버를 띄운 후 프론트에서 api에 접근할 때 서로 다른 포트일 경우 프록시 문제가 발생하였습니다.
백엔드 서버에 cors 설정을 하거나, 프론트 쪽에서 프록시 설정을 해주어 해당 포트에 정상적으로 접근 가능하게 만들 수 있는데.
일단 저 방법부터 사용해봤습니다!

crongro replied:
오 좋아요 잘 알고 썼군요. 프록시 문제라기보다 SOP문제라고 할 수 있어요.

crongro reviewed (approved):
width 로 progressbar늘리는것 말고 다른 방법도 생각해보세요.

wan0514 committed:
♻️ refactor: Remove magic numbers and clean up indentation

- Removed magic numbers and replaced them with named constants for better readability
- Cleaned up function indentation to follow consistent coding standards
- Minor refactoring to improve code clarity and maintainability

wan0514 committed:
♻️ refactor: Rename modal component props for clarity and reusability

- Renamed modal component props to more descriptive and generic names
- Improved prop names for better readability and future scalability

wan0514 committed:
♻️ refactor: Rename button component prop to avoid confusion with HTML attribute

- Changed the prop name to avoid conflict with the HTML 'label' attribute
- Ensured that non-string values won't cause errors by using a more specific prop name

wan0514 committed:
✨ feat: Implement subscribe and unsubscribe functionality in grid

- Added subscribe and unsubscribe functionality for media companies in the grid
- Identified an issue where pagination is not reset when switching between subscribe and all view
- This issue will be fixed in a future update

wan0514 committed:
🐛 fix: Fix pagination issue when switching between subscribe and all view

- Resolved the issue where pagination was not resetting when switching between subscribe and all view
- Refactored functions to pure functions for better readability and maintainability
- Plan to extract the functions into a utility file in the future

wan0514 committed:
♻️ refactor: Move utility functions to util folder

- Moved utility functions to a dedicated 'util' folder for better organization

wan0514 committed:
✨ feat: Dynamically calculate totalPage with a maximum limit of 4 pages

- Implemented dynamic calculation of totalPage, ensuring the maximum value doesn't exceed 4 pages
- Used `Math.min` to restrict the totalPage to a maximum of 4 while considering the filtered data length

wan0514 committed:
✨ feat: Implement useModal hook for global management of modal and its state using Context API

- Created a useModal hook to manage modal and modal state globally using Context API.
- Refactored previous code where modal-related state was scattered, now centralized for better consistency and maintainability.

wan0514 committed:
🐛 fix: Resolve bug where media section height disappears when grid cells are empty

- Fixed the issue where the height of the media section was disappearing when there were no grid cells.
- Adjusted CSS to ensure consistent layout even when grid is empty.

crongro reviewed (approved):
수고하셨어요.마지막 피드백까지 반영했군요 ㅎ

wan0514 committed:
✨ feat : grid subscribed

wan0514 committed:
Update README.md

wan0514 committed:
Update README.md
